bgpd/bgp_orr.c

```diff
@@ -0,0 +1,1163 @@
+					    safi_t safi, const char *name)
+{
+	return peer_af_flag_check(peer, afi, safi, PEER_FLAG_ORR_GROUP) &&
+	       !strcmp(peer->orr_group_name[afi][safi], name);
+}
+
+static inline bool is_peer_reachable(struct peer *peer, afi_t afi, safi_t safi)
+{
+	return peer && peer->afc_nego[afi][safi] && peer->status == Established;
+}
+
+static inline bool is_peer_active_eligible(struct peer *peer, afi_t afi,
+					   safi_t safi, const char *name)
+{
+	return is_peer_reachable(peer, afi, afi) &&
+	       is_peer_orr_group_member(peer, afi, safi, name);
+}
+
+static void bgp_orr_igp_metric_register(struct bgp_orr_group *orr_group,
+					bool reg);
+
+static void
+bgp_peer_update_orr_group_active_root(struct peer *peer, afi_t afi, safi_t safi,
+				      struct bgp_orr_group *orr_group);
+
+static struct bgp_orr_group *bgp_orr_group_new(struct bgp *bgp, afi_t afi,
+					       safi_t safi, const char *name)
+{
+	int ret;
+	struct list *orr_group_list = NULL;
+	struct bgp_orr_group *orr_group = NULL;
+
+	assert(bgp && name);
+
+	if (!bgp->orr_group[afi][safi])
+		bgp->orr_group[afi][safi] = list_new();
+
+	orr_group_list = bgp->orr_group[afi][safi];
+	orr_group = XCALLOC(MTYPE_BGP_ORR_GROUP, sizeof(struct bgp_orr_group));
+	assert(orr_group);
+
+	listnode_add(orr_group_list, orr_group);
+
+	orr_group->name = XSTRDUP(MTYPE_BGP_ORR_GROUP_NAME, name);
+	orr_group->afi = afi;
+	orr_group->safi = safi;
+	orr_group->primary = orr_group->secondary = orr_group->tertiary = NULL;
+	orr_group->bgp = bgp;
+
+	/* Initialize ORR Group route table */
+	orr_group->route_table = bgp_table_init(bgp, afi, safi);
+	assert(orr_group->route_table);
+
+	/* Register for opaque messages from IGPs when first ORR group is
+	 * configured. */
+	if (!bgp->orr_group_count) {
+		ret = zclient_register_opaque(zclient, ORR_IGP_METRIC_UPDATE);
+		if (ret != ZCLIENT_SEND_SUCCESS)
+			zlog_debug(
+				"%s: zclient_register_opaque failed with ret = %d",
+				__func__, ret);
+	}
+
+	bgp->orr_group_count++;
+
+	return orr_group;
+}
+
+static void bgp_orr_group_free(struct bgp_orr_group *orr_group)
+{
+	afi_t afi;
+	safi_t safi;
+	struct bgp *bgp;
+
+	assert(orr_group && orr_group->bgp && orr_group->name);
+
+	bgp_orr_debug("%s: Deleting ORR group %s", __func__, orr_group->name);
+
+	afi = orr_group->afi;
+	safi = orr_group->safi;
+	bgp = orr_group->bgp;
+
+	/* Unregister with IGP for metric calculation from specified location
+	 * and delete igp_metric_info calculated for this group */
+	bgp_orr_igp_metric_register(orr_group, false);
+
+	/* Free RR client list associated with this ORR group */
+	if (orr_group->rr_client_list)
+		list_delete(&orr_group->rr_client_list);
+
+	/* Free route table */
+	bgp_table_unlock(orr_group->route_table);
+	orr_group->route_table = NULL;
+
+	/* Unset ORR Group parameters */
+	XFREE(MTYPE_BGP_ORR_GROUP_NAME, orr_group->name);
+	memset(orr_group, 0, sizeof(struct bgp_orr_group));
+
+	listnode_delete(bgp->orr_group[afi][safi], orr_group);
```

Comment:
Previously it's zeroed, is this a valid call?

Reply:
addressed this comment. removed memset before listnode_delete